src/Handler/Workspace/Create/CreateWorkspaceHandler.php

```diff
@@ -0,0 +1,118 @@
+     */
+    public function __invoke(
+        Message $credentials,
+        Message $command,
+        array $features
+    ): ?Message {
+        assert($credentials instanceof GenericBackendCredentials);
+        assert($command instanceof CreateWorkspaceCommand);
+
+        // validate
+        assert(!empty($command->getProjectId()), 'CreateWorkspaceCommand.projectId is required');
+        assert(!empty($command->getWorkspaceId()), 'CreateWorkspaceCommand.workspaceId is required');
+        assert(!empty($command->getProjectUserName()), 'CreateWorkspaceCommand.projectUserName is required');
+        assert(!empty($command->getProjectRoleName()), 'CreateWorkspaceCommand.projectRoleName is required');
+        assert(!empty($command->getProjectReadOnlyRoleName()), 'CreateWorkspaceCommand.projectReadOnlyRoleName is required');
+
+        $db = $this->manager->createSession($credentials);
+        try {
+            // root user is also root database
+            $databaseName = $credentials->getPrincipal();
+
+            // TODO jaky stackPrefix?
+            $nameGenerator = NameGeneratorFactory::getGeneratorForBackendAndPrefix(
+                BackendSupportsInterface::BACKEND_TERADATA,
+                'workspace_'
+            );
+            // TODO pouzit projectId
+            $newWsDatabaseName = $nameGenerator->createWorkspaceObjectNameForWorkspaceId($command->getWorkspaceId());
+            $newWsUserName = $nameGenerator->createWorkspaceUserNameForWorkspaceId($command->getWorkspaceId());
+            $newWsRoleName = $nameGenerator->createWorkspaceRoleNameForWorkspaceId($command->getWorkspaceId());
+            $newWsPassword = Password::generate(
+                30,
+                Password::SET_LOWERCASE | Password::SET_UPPERCASE | Password::SET_NUMBER | Password::SET_SPECIAL_CHARACTERS
+            );
+
+            $db->executeStatement(sprintf(
+                'CREATE ROLE %s;',
+                TeradataQuote::quoteSingleIdentifier($newWsRoleName)
+            ));
+
+            // grant workspace role to root user
+            $db->executeStatement(sprintf(
+                'GRANT %s TO %s;',
+                TeradataQuote::quoteSingleIdentifier($newWsRoleName),
+                TeradataQuote::quoteSingleIdentifier($credentials->getPrincipal()),
+            ));
+
+            // grant workspace role to project user
+            $db->executeStatement(sprintf(
+                'GRANT %s TO %s;',
+                TeradataQuote::quoteSingleIdentifier($newWsRoleName),
+                TeradataQuote::quoteSingleIdentifier($command->getProjectRoleName()),
+            ));
+
+            // grant workspace role to project read only user
+            $db->executeStatement(sprintf(
+                'GRANT SELECT ON %s TO %s;',
+                TeradataQuote::quoteSingleIdentifier($newWsRoleName),
+                TeradataQuote::quoteSingleIdentifier($command->getProjectReadOnlyRoleName()),
+            ));
```

Comment:
To by mělo byt naopak ne? Chceš aby WS role měla práva číst přes readonly roli

Reply:
to je rekl bych spravne: moznost provadet select nad databazi pro uzivatele/roli.

Reply:
pridany testy na kontrolu prav